File: 2ndSemester/kursprogramowania/lista3/cpp/test/testRectangle.cpp
#include <stdexcept>
#include <cassert>
#include <cmath>

#include <rectangle.hpp>

void testRectangle() noexcept(true);
static void testRectangleConstructorExceptions(double side1, double side2) noexcept(true);
static void testRectangleConstructor() noexcept(true);

static void testRectangleConstructorExceptions(double side1, double side2) noexcept(true)
{
    try
    {
        const auto val = Rectangle(side1, side2);
        (void)val;

        assert(false && "Exception should be raised");
    }
    catch(const std::invalid_argument& e)
    {
        (void)e;
        assert(true);
    }
}

static void testRectangleConstructor() noexcept(true)
{
    {
        testRectangleConstructorExceptions(-100, 100);
        testRectangleConstructorExceptions(123, -123);
        testRectangleConstructorExceptions(-1, -1);
        testRectangleConstructorExceptions(0, 1);
        testRectangleConstructorExceptions(100, 0);
        testRectangleConstructorExceptions(-12345678, -12345678);

    }
    {
        Rectangle testRect = Rectangle(1, 1);
        assert(std::fabs(testRect.area() - 1) < 0.00001);
        assert(std::fabs(testRect.perimeter() - 4) < 0.00001);
        assert(testRect.name().compare("Rectangle") == 0);
    }
    {
        Rectangle testRect = Rectangle(5, 4);
        assert(std::fabs(testRect.area() - 20) < 0.00001);
        assert(std::fabs(testRect.perimeter() - 18) < 0.00001);
        assert(testRect.name().compare("Rectangle") == 0);
    }
    {
        Rectangle testRect = Rectangle(0.05, 0.2);
        assert(std::fabs(testRect.area() - 0.01) < 0.00001);
        assert(std::fabs(testRect.perimeter() - 0.5) < 0.00001);
        assert(testRect.name().compare("Rectangle") == 0);
    }
    {
        Rectangle testRect = Rectangle(240, 0.13);
        assert(std::fabs(testRect.area() - 31.2) < 0.00001);
        assert(std::fabs(testRect.perimeter() - 480.26) < 0.00001);
        assert(testRect.name().compare("Rectangle") == 0);
    }
}

void testRectangle() noexcept(true)
{
    testRectangleConstructor();
}

File: 2ndSemester/kursprogramowania/lista3/java/TestRectangle.java
public final class TestRectangle 
{
    private TestRectangle() throws InstantiationError
    {
        throw new InstantiationError("This is a static class");
    }    

    static private void testRectangleExceptions(final double side1, final double side2)
    {
        try
        {
            new Rectangle(side1, side2);
            assert false: "Exception should be raised.";
        }
        catch(final IllegalArgumentException e)
        {

        }
    }    

    static public void testRectangle()
    {
        {
            testRectangleExceptions(-100, 100);
            testRectangleExceptions(123, -123);
            testRectangleExceptions(-1, -1);
            testRectangleExceptions(0, 1);
            testRectangleExceptions(100, 0);
            testRectangleExceptions(-12345678, -12345678);
            testRectangleExceptions(-20.48, 40.96);
    
        }
        {
            Rectangle testRect = new Rectangle(1, 1);
            assert(Math.abs(testRect.area() - 1) < 0.00001);
            assert(Math.abs(testRect.perimeter() - 4) < 0.00001);
            assert(testRect.name() == "Rectangle");
        }
        {
            Rectangle testRect = new Rectangle(5, 4);
            assert(Math.abs(testRect.area() - 20) < 0.00001);
            assert(Math.abs(testRect.perimeter() - 18) < 0.00001);
            assert(testRect.name() == "Rectangle");
        }
        {
            Rectangle testRect = new Rectangle(0.05, 0.2);
            assert(Math.abs(testRect.area() - 0.01) < 0.00001);
            assert(Math.abs(testRect.perimeter() - 0.5) < 0.00001);
            assert(testRect.name() == "Rectangle");
        }
        {
            Rectangle testRect = new Rectangle(240, 0.13);
            assert(Math.abs(testRect.area() - 31.2) < 0.00001);
            assert(Math.abs(testRect.perimeter() - 480.26) < 0.00001);
            assert(testRect.name() == "Rectangle");
        }
    }
}
